apollo-federation/src/query_plan/operation.rs

```diff
@@ -4125,12 +4130,13 @@ impl InlineFragmentSelection {
                         }),
                         None,
                     );
-
+                    // Return `... on <rebased condition> { __typename @include(if: false) }`
+                    let rebased_casted_type = rebased_fragment.data().casted_type();
                     return Ok(Some(SelectionOrSet::Selection(
                         Selection::from_inline_fragment(
                             rebased_fragment,
                             SelectionSet::from_selection(
-                                parent_type.clone(),
+                                rebased_casted_type,
```

Comment:
Hmmm! we should make `Selection::from_inline_fragment` should enforce this. The SelectionSet must always have the fragment's type condition as its parent type (or the fragment's parent type if it does not have a type condition).

Reply:
I don't think this is necessary true as inline fragment without type condition is still valid i.e. `.... { x y }` is a valid (useless) fragment.

Reply:
*this was one of the issues i hit with debugging `handle_requires` tests

Reply:
Yes, then the selection type must match the fragment's parent type.

Reply:
inline fragment's sub-selection's type must be the same as its casted type. I'll add this assertion.